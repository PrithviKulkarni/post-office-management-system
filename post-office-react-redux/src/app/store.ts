import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
