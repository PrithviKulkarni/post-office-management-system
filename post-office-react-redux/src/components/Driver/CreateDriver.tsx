import { useState, SyntheticEvent } from "react";
import { useNavigate } from "react-router-dom";
import { Container, Button, Form, InputGroup } from "react-bootstrap";
import { Driver, defaultDriver } from "../interfaces/driverinterface";
import { addNewDriver } from "../../features/DriverSlice";
import { useAppDispatch } from "../../app/hooks";

const CreateDriver = () => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const [fname, setFirstname] = useState<string>("");
  const [lname, setLastname] = useState<string>("");
  const [username, setUsername] = useState<string>("");
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [fnameVal, setFirstnameVal] = useState<boolean>(false);
  const [lnameVal, setLastnameVal] = useState<boolean>(false);
  const [usernameVal, setUsernameVal] = useState<boolean>(false);
  const [emailVal, setEmailVal] = useState<boolean>(false);
  const [passwordVal, setPasswordVal] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");

  /**
   * *
   * handles the submit for confirmation and sends a post or patch request to the mongoDB.
   *
   *  @returns post request Drivers. */

  const submitHandler = (event: SyntheticEvent) => {
    event.preventDefault();

    const validated = [fname, lname, username, email, password].every(Boolean);
    try {
      if (!fname) {
        setFirstnameVal(true);
      } else {
        setFirstnameVal(false);
      }
      if (!lname) {
        setLastnameVal(true);
      } else {
        setLastnameVal(false);
      }
      if (!username) {
        setUsernameVal(true);
      } else {
        setUsernameVal(false);
      }
      if (!email) {
        setEmailVal(true);
      } else {
        setEmailVal(false);
      }
      if (!password) {
        setPasswordVal(true);
      } else {
        setPasswordVal(false);
      }

      if (validated) {
        const newDriver: Driver = defaultDriver;

        if (fname) {
          newDriver.first_name = fname;
        }
        if (lname) {
          newDriver.last_name = lname;
        }
        if (username) {
          newDriver.username = username;
        }
        if (email) {
          newDriver.email = email;
        }
        if (password) {
          newDriver.password = password;
        }

        console.log({ newDriver });

        dispatch(addNewDriver(newDriver));
        navigate("/drivers");
      }
    } catch (error) {
      setErrorMessage("Player NOT FOUND.");
      console.log("errorMessage: ", errorMessage);
    }
  };

  /**
   * *
   * Returns submitted information.
   *
   *  @returns Players Games. */

  return (
    <div className="create-driver" data-testid="create-driver-test">
      <Container>
        <Form
          onSubmit={submitHandler}
          style={{
            height: "100%",
            display: "flex",
            flexDirection: "column",
            justifyContent: "center",
            alignItems: "center",
            width: "100%",
          }}
          className="create-driver-form"
          data-testid="create-driver-form"
        >
          <InputGroup>
            <InputGroup.Text id="basic-addon1">First name</InputGroup.Text>
            <Form.Control
              type="text"
              className="fname"
              data-testid="create-driver-fname"
              aria-label="fname"
              aria-describedby="basic-addon1"
              size="sm"
              name="fname"
              value={fname}
              onChange={(event) => setFirstname(event.target.value)}
            />
            {fnameVal && (
              <Form.Text className="text-muted" data-testid="firstNameVal">
                Please fill in field
              </Form.Text>
            )}
          </InputGroup>

          <InputGroup>
            <InputGroup.Text id="basic-addon1">Last name</InputGroup.Text>
            <Form.Control
              type="text"
              className="lname"
              data-testid="create-driver-lname"
              aria-label="lname"
              aria-describedby="basic-addon1"
              size="sm"
              name="lname"
              value={lname}
              onChange={(event) => setLastname(event.target.value)}
            />
            {lnameVal && (
              <Form.Text className="text-muted" data-testid="lastNameVal">
                Please fill in field
              </Form.Text>
            )}
          </InputGroup>

          <InputGroup>
            <InputGroup.Text id="basic-addon1">Username</InputGroup.Text>
            <Form.Control
              type="text"
              aria-label="username"
              className="username"
              data-testid="create-driver-username"
              aria-describedby="basic-addon1"
              size="sm"
              name="username"
              value={username}
              onChange={(event) => setUsername(event.target.value)}
            />
            {usernameVal && (
              <Form.Text className="text-muted" data-testid="usernameVal">
                Please fill in field
              </Form.Text>
            )}
          </InputGroup>

          <InputGroup>
            <InputGroup.Text id="basic-addon1">Email</InputGroup.Text>
            <Form.Control
              type="text"
              aria-label="email"
              className="email"
              data-testid="create-driver-email"
              aria-describedby="basic-addon1"
              size="sm"
              name="email"
              value={email}
              onChange={(event) => setEmail(event.target.value)}
            />
            {emailVal && (
              <Form.Text className="text-muted" data-testid="emailVal">
                Please fill in field
              </Form.Text>
            )}
          </InputGroup>

          <InputGroup>
            <InputGroup.Text id="basic-addon1">Password</InputGroup.Text>
            <Form.Control
              type="text"
              aria-label="password"
              className="password"
              data-testid="create-driver-password"
              aria-describedby="basic-addon1"
              size="sm"
              name="password"
              value={password}
              onChange={(event) => setPassword(event.target.value)}
            />
            {passwordVal && (
              <Form.Text className="text-muted" data-testid="passwordVal">
                Please fill in field
              </Form.Text>
            )}
          </InputGroup>
          <Button type="submit" data-testid="driver-create-button">
            Submit
          </Button>
          <p>{errorMessage}</p>
        </Form>
      </Container>
    </div>
  );
};

export default CreateDriver;
