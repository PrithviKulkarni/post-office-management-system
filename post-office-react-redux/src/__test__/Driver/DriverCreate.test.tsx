import { render, cleanup, fireEvent } from '@testing-library/react';
import { screen } from '@testing-library/react';
import { Provider } from "react-redux";
import { store } from '../../app/store';
import { BrowserRouter } from "react-router-dom";
import CreateDriver from '../../components/Driver/CreateDriver'

afterEach(cleanup);

// CREATE DRIVERS COMPONENT

it('Renders Creat Drivers Page Correctly', () => {
     const { asFragment } = render(
          <Provider store={store}>
               <BrowserRouter>
                    <CreateDriver />
               </BrowserRouter>
          </Provider>
     );
     expect(asFragment()).toMatchSnapshot();
});

it('Should Verify Create Driver Render', () => {
     const { getByTestId } = render(
          <Provider store={store}>
               <BrowserRouter>
                    <CreateDriver />
               </BrowserRouter>
          </Provider>
     );
     const button = getByTestId("driver-create-button");
     expect(button.textContent).toBe("Submit");
     fireEvent.click(button);
     expect(button).toBeTruthy();

     const firstName = getByTestId("firstNameVal");
     expect(firstName.textContent).toBe("Please fill in field");
     const lastName = getByTestId("lastNameVal");
     expect(lastName.textContent).toBe("Please fill in field");
     const username = getByTestId("usernameVal");
     expect(username.textContent).toBe("Please fill in field");
     const email = getByTestId("emailVal");
     expect(email.textContent).toBe("Please fill in field");
     const password = getByTestId("passwordVal");
     expect(password.textContent).toBe("Please fill in field");
     expect(screen.getByTestId("create-driver-test")).toHaveClass("create-driver");

});

it('Should Verify Create Driver Render When Fields Are Filled', () => {
     const { getByTestId } = render(
          <Provider store={store}>
               <BrowserRouter>
                    <CreateDriver />
               </BrowserRouter>
          </Provider>
     );
     const button = getByTestId("driver-create-button");
     expect(button.textContent).toBe("Submit");
     const firstName = getByTestId("create-driver-fname");
     fireEvent.change(firstName, { target: { value: "Zainab" } });
     const lastName = getByTestId("create-driver-lname");
     fireEvent.change(lastName, { target: { value: "Bikar" } });
     const username = getByTestId("create-driver-username");
     fireEvent.change(username, { target: { value: "Zainab2022" } });
     const email = getByTestId("create-driver-email");
     fireEvent.change(email, { target: { value: "Zainab@gmail.com" } });
     const password = getByTestId("create-driver-password");
     fireEvent.change(password, { target: { value: "dfnrirje" } });
     fireEvent.click(button);
     expect(screen.getByTestId("create-driver-test")).toHaveClass("create-driver");
});
