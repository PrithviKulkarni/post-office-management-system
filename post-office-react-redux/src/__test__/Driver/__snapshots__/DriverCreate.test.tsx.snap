// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Creat Drivers Page Correctly 1`] = `
<DocumentFragment>
  <div
    class="create-driver"
    data-testid="create-driver-test"
  >
    <div
      class="container"
    >
      <form
        class="create-driver-form"
        data-testid="create-driver-form"
        style="height: 100%; display: flex; flex-direction: column; justify-content: center; align-items: center; width: 100%;"
      >
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            First name
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="fname"
            class="fname form-control form-control-sm"
            data-testid="create-driver-fname"
            name="fname"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Last name
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="lname"
            class="lname form-control form-control-sm"
            data-testid="create-driver-lname"
            name="lname"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Username
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="username"
            class="username form-control form-control-sm"
            data-testid="create-driver-username"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Email
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="email"
            class="email form-control form-control-sm"
            data-testid="create-driver-email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Password
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="password"
            class="password form-control form-control-sm"
            data-testid="create-driver-password"
            name="password"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          data-testid="driver-create-button"
          type="submit"
        >
          Submit
        </button>
        <p />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
